File: mobile/packages/divine_camera/android/src/main/kotlin/co/openvine/divine_camera/DivineCameraPlugin.kt
// ABOUTME: Main Flutter plugin entry point for Android camera operations
// ABOUTME: Handles method channel communication and delegates to CameraController

package co.openvine.divine_camera

import android.app.Activity
import android.content.Context
import android.media.AudioDeviceInfo
import android.media.AudioManager
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import io.flutter.view.TextureRegistry
import java.util.concurrent.atomic.AtomicBoolean

/** DivineCameraPlugin */
class DivineCameraPlugin :
    FlutterPlugin,
    MethodCallHandler,
    ActivityAware {

    private lateinit var channel: MethodChannel
    private lateinit var context: Context
    private lateinit var textureRegistry: TextureRegistry
    private var activity: Activity? = null
    private var cameraController: CameraController? = null
    private var volumeKeyHandler: VolumeKeyHandler? = null

    private val mainHandler = android.os.Handler(android.os.Looper.getMainLooper())

    // Per-instance forwarder that pushes native diagnostics over THIS engine's
    // channel. `DivineCameraLog.sink` is a process-wide singleton, so a second
    // FlutterEngine (e.g. the FCM background isolate that also runs
    // GeneratedPluginRegistrant) would otherwise overwrite it and route camera
    // logs to the wrong isolate. We re-assert this in onMethodCall — camera
    // calls only ever reach the UI engine — so the sink always points back here.
    private val logSink: (String, String, String) -> Unit = { level, message, name ->
        mainHandler.post {
            channel.invokeMethod(
                "onNativeLog",
                mapOf("level" to level, "message" to message, "name" to name)
            )
        }
    }

    override fun onAttachedToEngine(flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
        channel = MethodChannel(flutterPluginBinding.binaryMessenger, "divine_camera")
        channel.setMethodCallHandler(this)
        context = flutterPluginBinding.applicationContext
        textureRegistry = flutterPluginBinding.textureRegistry
        DivineCameraLog.sink = logSink
    }

    // Session-lifecycle operations worth leaving a breadcrumb for. High-
    // frequency calls (zoom / focus / exposure / getCameraState) are excluded
    // on purpose so they don't flood the captured log buffer.
    private val lifecycleMethods = setOf(
        "initializeCamera", "disposeCamera", "switchCamera",
        "startRecording", "stopRecording", "pausePreview", "resumePreview",
        "setFlashMode", "setRemoteRecordControlEnabled"
    )

    private fun logLifecycleCall(call: MethodCall) {
        if (call.method !in lifecycleMethods) return
        val args = call.arguments
        DivineCameraLog.debug(
            "→ ${call.method}${if (args != null) " $args" else ""}",
            name = "DivineCamera.Lifecycle"
        )
    }

    override fun onMethodCall(call: MethodCall, result: Result) {
        // Re-claim the shared sink in case another FlutterEngine overwrote it.
        DivineCameraLog.sink = logSink
        logLifecycleCall(call)
        val oneShotResult = OneShotMethodResult(result)
        when (call.method) {
            "getPlatformVersion" -> {
                oneShotResult.success("Android ${android.os.Build.VERSION.RELEASE}")
            }

            "initializeCamera" -> {
                val lens = call.argument<String>("lens") ?: "back"
                val videoQuality = call.argument<String>("videoQuality") ?: "fhd"
                val enableScreenFlash = call.argument<Boolean>("enableScreenFlash") ?: true
                val mirrorFrontCameraOutput = call.argument<Boolean>("mirrorFrontCameraOutput") ?: true
                val enableAutoLensSwitch = call.argument<Boolean>("enableAutoLensSwitch") ?: true
                initializeCamera(lens, videoQuality, enableScreenFlash, mirrorFrontCameraOutput, enableAutoLensSwitch, oneShotResult)
            }

            "disposeCamera" -> {
                disposeCamera(oneShotResult)
            }

            "setFlashMode" -> {
                val mode = call.argument<String>("mode") ?: "off"
                setFlashMode(mode, oneShotResult)
            }

            "setFocusPoint" -> {
                val x = call.argument<Double>("x") ?: 0.5
                val y = call.argument<Double>("y") ?: 0.5
                setFocusPoint(x.toFloat(), y.toFloat(), oneShotResult)
            }

            "setExposurePoint" -> {
                val x = call.argument<Double>("x") ?: 0.5
                val y = call.argument<Double>("y") ?: 0.5
                setExposurePoint(x.toFloat(), y.toFloat(), oneShotResult)
            }

            "cancelFocusAndMetering" -> {
                cancelFocusAndMetering(oneShotResult)
            }

            "setZoomLevel" -> {
                val level = call.argument<Double>("level") ?: 1.0
                setZoomLevel(level.toFloat(), oneShotResult)
            }

            "switchCamera" -> {
                val lens = call.argument<String>("lens") ?: "back"
                switchCamera(lens, oneShotResult)
            }

            "startRecording" -> {
                val maxDurationMs = call.argument<Int>("maxDurationMs")
                val useCache = call.argument<Boolean>("useCache") ?: true
                val outputDirectory = call.argument<String>("outputDirectory")
                startRecording(maxDurationMs, useCache, outputDirectory, oneShotResult)
            }

            "stopRecording" -> {
                stopRecording(oneShotResult)
            }

            "pausePreview" -> {
                pausePreview(oneShotResult)
            }

            "resumePreview" -> {
                resumePreview(oneShotResult)
            }

            "getCameraState" -> {
                getCameraState(oneShotResult)
            }

            "setRemoteRecordControlEnabled" -> {
                val enabled = call.argument<Boolean>("enabled") ?: false
                setRemoteRecordControlEnabled(enabled, oneShotResult)
            }

            "setVolumeKeysEnabled" -> {
                val enabled = call.argument<Boolean>("enabled") ?: true
                setVolumeKeysEnabled(enabled, oneShotResult)
            }

            "listAudioDevices" -> {
                listAudioDevices(result)
            }

            else -> {
                oneShotResult.notImplemented()
            }
        }
    }

    private fun initializeCamera(lens: String, videoQuality: String, enableScreenFlash: Boolean, mirrorFrontCameraOutput: Boolean, enableAutoLensSwitch: Boolean, result: Result) {
        val currentActivity = activity
        if (currentActivity == null) {
            result.error("NO_ACTIVITY", "Activity not available", null)
            return
        }

        try {
            cameraController?.release()
            cameraController = CameraController(
                context = context,
                activity = currentActivity,
                textureRegistry = textureRegistry
            )

            // Set up auto-stop listener to notify Flutter
            cameraController?.onAutoStopListener = { recordingResult ->
                channel.invokeMethod("onRecordingAutoStopped", recordingResult)
            }

            cameraController?.initialize(lens, videoQuality, enableScreenFlash, mirrorFrontCameraOutput, enableAutoLensSwitch) { state, error ->
                if (error != null) {
                    result.error("INIT_ERROR", error, null)
                } else {
                    val dict = state as? Map<*, *>
                    DivineCameraLog.info(
                        "Camera initialized (lens=${dict?.get("lens")}, " +
                            "aspectRatio=${dict?.get("aspectRatio")}, " +
                            "hasFlash=${dict?.get("hasFlash")}, " +
                            "lenses=${dict?.get("availableLenses")})",
                        name = "DivineCamera.Lifecycle"
                    )
                    result.success(state)
                }
            }
        } catch (e: Exception) {
            result.error("INIT_EXCEPTION", e.message, e.stackTraceToString())
        }
    }

    private fun disposeCamera(result: Result) {
        try {
            volumeKeyHandler?.release()
            volumeKeyHandler = null
            cameraController?.release()
            cameraController = null
            result.success(null)
        } catch (e: Exception) {
            result.error("DISPOSE_ERROR", e.message, null)
        }
    }

    private fun setFlashMode(mode: String, result: Result) {
        val controller = cameraController
        if (controller == null) {
            result.error("NOT_INITIALIZED", "Camera not initialized", null)
            return
        }
        try {
            val success = controller.setFlashMode(mode)
            result.success(success)
        } catch (e: Exception) {
            result.error("FLASH_ERROR", e.message, null)
        }
    }

    private fun setFocusPoint(x: Float, y: Float, result: Result) {
        val controller = cameraController
        if (controller == null) {
            result.error("NOT_INITIALIZED", "Camera not initialized", null)
            return
        }
        try {
            val success = controller.setFocusPoint(x, y)
            result.success(success)
        } catch (e: Exception) {
            result.error("FOCUS_ERROR", e.message, null)
        }
    }

    private fun setExposurePoint(x: Float, y: Float, result: Result) {
        val controller = cameraController
        if (controller == null) {
            result.error("NOT_INITIALIZED", "Camera not initialized", null)
            return
        }
        try {
            val success = controller.setExposurePoint(x, y)
            result.success(success)
        } catch (e: Exception) {
            result.error("EXPOSURE_ERROR", e.message, null)
        }
    }

    private fun cancelFocusAndMetering(result: Result) {
        val controller = cameraController
        if (controller == null) {
            result.error("NOT_INITIALIZED", "Camera not initialized", null)
            return
        }
        try {
            val success = controller.cancelFocusAndMetering()
            result.success(success)
        } catch (e: Exception) {
            result.error("FOCUS_ERROR", e.message, null)
        }
    }

    private fun setZoomLevel(level: Float, result: Result) {
        val controller = cameraController
        if (controller == null) {
            result.error("NOT_INITIALIZED", "Camera not initialized", null)
            return
        }
        try {
            val success = controller.setZoomLevel(level)
            result.success(success)
        } catch (e: Exception) {
            result.error("ZOOM_ERROR", e.message, null)
        }
    }

    private fun switchCamera(lens: String, result: Result) {
        val controller = cameraController
        if (controller == null) {
            result.error("NOT_INITIALIZED", "Camera not initialized", null)
            return
        }
        try {
            // Suppress volume/Bluetooth triggers during camera switch.
            // Camera reconfiguration can cause connected Bluetooth devices
            // to send spurious play/pause events that would start recording.
            volumeKeyHandler?.suppressTemporarily(3000)

            controller.switchCamera(lens) { state, error ->
                if (error != null) {
                    result.error("SWITCH_ERROR", error, null)
                } else {
                    val dict = state as? Map<*, *>
                    DivineCameraLog.info(
                        "Camera switched (lens=${dict?.get("lens")})",
                        name = "DivineCamera.Lifecycle"
                    )
                    result.success(state)
                }
            }
        } catch (e: Exception) {
            result.error("SWITCH_EXCEPTION", e.message, null)
        }
    }

    private fun startRecording(maxDurationMs: Int?, useCache: Boolean, outputDirectory: String?, result: Result) {
        val controller = cameraController
        if (controller == null) {
            result.error("NOT_INITIALIZED", "Camera not initialized", null)
            return
        }
        try {
            controller.startRecording(maxDurationMs, useCache, outputDirectory) { error ->
                if (error != null) {
                    result.error("RECORD_START_ERROR", error, null)
                } else {
                    result.success(null)
                }
            }
        } catch (e: Exception) {
            result.error("RECORD_START_EXCEPTION", e.message, null)
        }
    }

    private fun stopRecording(result: Result) {
        val controller = cameraController
        if (controller == null) {
            result.error("NOT_INITIALIZED", "Camera not initialized", null)
            return
        }
        try {
            controller.stopRecording { recordingResult, error ->
                if (error != null) {
                    result.error("RECORD_STOP_ERROR", error, null)
                } else {
                    result.success(recordingResult)
                }
            }
        } catch (e: Exception) {
            result.error("RECORD_STOP_EXCEPTION", e.message, null)
        }
    }

    private fun pausePreview(result: Result) {
        try {
            cameraController?.pausePreview()
            result.success(null)
        } catch (e: Exception) {
            result.error("PAUSE_ERROR", e.message, null)
        }
    }

    private fun resumePreview(result: Result) {
        try {
            cameraController?.resumePreview { state, error ->
                if (error != null) {
                    result.error("RESUME_ERROR", error, null)
                } else {
                    result.success(state)
                }
            }
        } catch (e: Exception) {
            result.error("RESUME_EXCEPTION", e.message, null)
        }
    }

    private fun getCameraState(result: Result) {
        val controller = cameraController
        if (controller == null) {
            result.error("NOT_INITIALIZED", "Camera not initialized", null)
            return
        }
        try {
            result.success(controller.getCameraState())
        } catch (e: Exception) {
            result.error("STATE_ERROR", e.message, null)
        }
    }

    private fun setRemoteRecordControlEnabled(enabled: Boolean, result: Result) {
        try {
            if (enabled) {
                if (volumeKeyHandler == null) {
                    volumeKeyHandler = VolumeKeyHandler(context, activity) { triggerType ->
                        // Send trigger event to Flutter on main thread
                        android.os.Handler(android.os.Looper.getMainLooper()).post {
                            channel.invokeMethod("onRemoteRecordTrigger", triggerType)
                        }
                    }
                }
                val success = volumeKeyHandler?.enable() ?: false
                result.success(success)
            } else {
                volumeKeyHandler?.disable()
                result.success(true)
            }
        } catch (e: Exception) {
            result.error("REMOTE_CONTROL_ERROR", e.message, null)
        }
    }

    private fun setVolumeKeysEnabled(enabled: Boolean, result: Result) {
        try {
            volumeKeyHandler?.setVolumeKeysEnabled(enabled)
            result.success(true)
        } catch (e: Exception) {
            result.error("VOLUME_KEYS_ERROR", e.message, null)
        }
    }

    private fun listAudioDevices(result: Result) {
        val audioManager = context?.getSystemService(Context.AUDIO_SERVICE) as? AudioManager
        if (audioManager == null) {
            result.success(emptyList<Map<String, String>>())
            return
        }
        val devices = audioManager.getDevices(AudioManager.GET_DEVICES_INPUTS)
            .filter { it.type != AudioDeviceInfo.TYPE_TELEPHONY }
            .map { device ->
                mapOf(
                    "id" to device.id.toString(),
                    "name" to (device.productName?.toString() ?: "Audio Device ${device.id}")
                )
            }
        result.success(devices)
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel.setMethodCallHandler(null)
        // Only relinquish the shared sink if it still points at this instance,
        // so a background-engine teardown can't silence the UI engine's logs.
        if (DivineCameraLog.sink === logSink) {
            DivineCameraLog.sink = null
        }
        volumeKeyHandler?.release()
        volumeKeyHandler = null
        cameraController?.release()
        cameraController = null
    }

    override fun onAttachedToActivity(binding: ActivityPluginBinding) {
        activity = binding.activity
    }

    override fun onDetachedFromActivityForConfigChanges() {
        activity = null
    }

    override fun onReattachedToActivityForConfigChanges(binding: ActivityPluginBinding) {
        activity = binding.activity
    }

    override fun onDetachedFromActivity() {
        activity = null
    }
}

internal class OneShotMethodResult(
    private val delegate: Result
) : Result {
    private val replied = AtomicBoolean(false)

    override fun success(result: Any?) {
        if (markReplied()) {
            delegate.success(result)
        }
    }

    override fun error(errorCode: String, errorMessage: String?, errorDetails: Any?) {
        if (markReplied()) {
            // Choke point for every error returned to Dart — forward it to the
            // UnifiedLogger so the failure shows up in user bug reports. Fatal
            // native crashes are captured separately by Crashlytics.
            DivineCameraLog.error(
                "$errorCode: ${errorMessage ?: ""}",
                name = "DivineCamera.Plugin"
            )
            delegate.error(errorCode, errorMessage, errorDetails)
        }
    }

    override fun notImplemented() {
        if (markReplied()) {
            delegate.notImplemented()
        }
    }

    private fun markReplied(): Boolean = replied.compareAndSet(false, true)
}

File: mobile/packages/divine_camera/macos/Classes/DivineCameraPlugin.swift
// ABOUTME: Main Flutter plugin entry point for macOS camera operations
// ABOUTME: Handles method channel communication and delegates to CameraController

import FlutterMacOS
import AppKit
import AVFoundation

public class DivineCameraPlugin: NSObject, FlutterPlugin {
    private var cameraController: CameraController?
    private var textureRegistry: FlutterTextureRegistry?
    private var messenger: FlutterBinaryMessenger?
    private var methodChannel: FlutterMethodChannel?

    public static func register(with registrar: FlutterPluginRegistrar) {
        let channel = FlutterMethodChannel(
            name: "divine_camera",
            binaryMessenger: registrar.messenger
        )
        let instance = DivineCameraPlugin()
        instance.textureRegistry = registrar.textures
        instance.messenger = registrar.messenger
        instance.methodChannel = channel
        registrar.addMethodCallDelegate(instance, channel: channel)

        // Forward curated native diagnostics to Dart's UnifiedLogger so
        // recording issues (e.g. missing audio) land in user bug reports.
        instance.installLogSink()

        // Listen for auto-stop events from CameraController
        NotificationCenter.default.addObserver(
            instance,
            selector: #selector(instance.handleAutoStop(_:)),
            name: NSNotification.Name("DivineCameraAutoStop"),
            object: nil
        )
    }

    @objc private func handleAutoStop(_ notification: Notification) {
        guard let userInfo = notification.userInfo as? [String: Any] else { return }

        DispatchQueue.main.async { [weak self] in
            self?.methodChannel?.invokeMethod(
                "onRecordingAutoStopped",
                arguments: userInfo
            )
        }
    }

    /// Session-lifecycle operations worth leaving a breadcrumb for. High-
    /// frequency calls (zoom / focus / exposure / getCameraState) are excluded
    /// on purpose so they don't flood the captured log buffer.
    private static let lifecycleMethods: Set<String> = [
        "initializeCamera", "disposeCamera", "switchCamera",
        "startRecording", "stopRecording", "pausePreview", "resumePreview",
        "setFlashMode",
    ]

    private static func logLifecycleCall(_ call: FlutterMethodCall) {
        guard lifecycleMethods.contains(call.method) else { return }
        let args = call.arguments as? [String: Any]
        DivineCameraLog.shared.debug(
            "→ \(call.method)\(args.map { " \($0)" } ?? "")",
            name: "DivineCamera.Lifecycle"
        )
    }

    /// Per-instance forwarder pushing native diagnostics over THIS engine's
    /// channel. `DivineCameraLog.shared.sink` is a process-wide singleton, so a
    /// second FlutterEngine that registers the plugin would otherwise overwrite
    /// it and route camera logs to the wrong isolate. We re-assert it in
    /// `handle` — camera calls only ever reach the UI engine.
    private lazy var logSink: (String, String, String) -> Void = {
        [weak self] level, message, name in
        DispatchQueue.main.async {
            self?.methodChannel?.invokeMethod(
                "onNativeLog",
                arguments: ["level": level, "message": message, "name": name]
            )
        }
    }

    private func installLogSink() {
        DivineCameraLog.shared.sink = logSink
    }

    public func handle(
        _ call: FlutterMethodCall,
        result: @escaping FlutterResult
    ) {
        installLogSink()
        DivineCameraPlugin.logLifecycleCall(call)
        switch call.method {
        case "getPlatformVersion":
            result(
                "macOS " + ProcessInfo.processInfo.operatingSystemVersionString
            )

        case "initializeCamera":
            let args = call.arguments as? [String: Any] ?? [:]
            let lens = args["lens"] as? String ?? "front"
            let videoQuality = args["videoQuality"] as? String ?? "fhd"
            let enableAutoLensSwitch =
                args["enableAutoLensSwitch"] as? Bool ?? false
            initializeCamera(
                lens: lens,
                videoQuality: videoQuality,
                enableAutoLensSwitch: enableAutoLensSwitch,
                result: result
            )

        case "disposeCamera":
            disposeCamera(result: result)

        case "setFlashMode":
            let args = call.arguments as? [String: Any] ?? [:]
            let mode = args["mode"] as? String ?? "off"
            setFlashMode(mode: mode, result: result)

        case "setFocusPoint":
            let args = call.arguments as? [String: Any] ?? [:]
            let x = args["x"] as? Double ?? 0.5
            let y = args["y"] as? Double ?? 0.5
            setFocusPoint(x: x, y: y, result: result)

        case "setExposurePoint":
            let args = call.arguments as? [String: Any] ?? [:]
            let x = args["x"] as? Double ?? 0.5
            let y = args["y"] as? Double ?? 0.5
            setExposurePoint(x: x, y: y, result: result)

        case "cancelFocusAndMetering":
            cancelFocusAndMetering(result: result)

        case "setZoomLevel":
            let args = call.arguments as? [String: Any] ?? [:]
            let level = args["level"] as? Double ?? 1.0
            setZoomLevel(level: level, result: result)

        case "switchCamera":
            let args = call.arguments as? [String: Any] ?? [:]
            let lens = args["lens"] as? String ?? "front"
            switchCamera(lens: lens, result: result)

        case "startRecording":
            let args = call.arguments as? [String: Any] ?? [:]
            let maxDurationMs = args["maxDurationMs"] as? Int
            let useCache = args["useCache"] as? Bool ?? true
            let outputDirectory = args["outputDirectory"] as? String
            let audioDeviceId = args["audioDeviceId"] as? String
            startRecording(
                maxDurationMs: maxDurationMs,
                useCache: useCache,
                outputDirectory: outputDirectory,
                audioDeviceId: audioDeviceId,
                result: result
            )

        case "stopRecording":
            stopRecording(result: result)

        case "pausePreview":
            pausePreview(result: result)

        case "resumePreview":
            resumePreview(result: result)

        case "getCameraState":
            getCameraState(result: result)

        case "setRemoteRecordControlEnabled":
            // Remote record control (volume keys / Bluetooth) not supported
            // on macOS
            result(false)

        case "setVolumeKeysEnabled":
            // Volume keys not supported on macOS
            result(false)

        case "listAudioDevices":
            listAudioDevices(result: result)

        default:
            result(FlutterMethodNotImplemented)
        }
    }

    // MARK: - Camera Operations

    /// Builds a `FlutterError` and forwards it to Dart's UnifiedLogger so the
    /// failure shows up in user bug reports. Fatal native crashes are captured
    /// separately by Crashlytics. Static so it can be called from `@escaping`
    /// completion closures without capturing `self`.
    private static func cameraError(_ code: String, _ message: String?) -> FlutterError {
        DivineCameraLog.shared.error(
            "\(code): \(message ?? "")",
            name: "DivineCamera.Plugin"
        )
        return FlutterError(code: code, message: message, details: nil)
    }

    private func initializeCamera(
        lens: String,
        videoQuality: String,
        enableAutoLensSwitch: Bool,
        result: @escaping FlutterResult
    ) {
        guard let registry = textureRegistry else {
            result(Self.cameraError("NO_REGISTRY", "Texture registry not available"))
            return
        }

        cameraController?.release()
        cameraController = CameraController(textureRegistry: registry)

        cameraController?.initialize(
            lens: lens,
            videoQuality: videoQuality,
            enableAutoLensSwitch: enableAutoLensSwitch
        ) { state, error in
            DispatchQueue.main.async {
                if let error = error {
                    result(Self.cameraError("INIT_ERROR", error))
                } else {
                    if let dict = state as? [String: Any] {
                        DivineCameraLog.shared.info(
                            "Camera initialized (lens=\(dict["lens"] ?? "?"), "
                                + "aspectRatio=\(dict["aspectRatio"] ?? "?"), "
                                + "hasFlash=\(dict["hasFlash"] ?? "?"), "
                                + "lenses=\(dict["availableLenses"] ?? "?"))",
                            name: "DivineCamera.Lifecycle"
                        )
                    }
                    result(state)
                }
            }
        }
    }

    private func disposeCamera(result: @escaping FlutterResult) {
        cameraController?.release()
        cameraController = nil
        result(nil)
    }

    private func setFlashMode(
        mode: String,
        result: @escaping FlutterResult
    ) {
        guard let controller = cameraController else {
            result(Self.cameraError("NOT_INITIALIZED", "Camera not initialized"))
            return
        }
        let success = controller.setFlashMode(mode: mode)
        result(success)
    }

    private func setFocusPoint(
        x: Double,
        y: Double,
        result: @escaping FlutterResult
    ) {
        guard let controller = cameraController else {
            result(Self.cameraError("NOT_INITIALIZED", "Camera not initialized"))
            return
        }
        let success = controller.setFocusPoint(x: CGFloat(x), y: CGFloat(y))
        result(success)
    }

    private func setExposurePoint(
        x: Double,
        y: Double,
        result: @escaping FlutterResult
    ) {
        guard let controller = cameraController else {
            result(Self.cameraError("NOT_INITIALIZED", "Camera not initialized"))
            return
        }
        let success = controller.setExposurePoint(
            x: CGFloat(x),
            y: CGFloat(y)
        )
        result(success)
    }

    private func cancelFocusAndMetering(result: @escaping FlutterResult) {
        guard let controller = cameraController else {
            result(Self.cameraError("NOT_INITIALIZED", "Camera not initialized"))
            return
        }
        let success = controller.cancelFocusAndMetering()
        result(success)
    }

    private func setZoomLevel(level: Double, result: @escaping FlutterResult) {
        guard let controller = cameraController else {
            result(Self.cameraError("NOT_INITIALIZED", "Camera not initialized"))
            return
        }
        let success = controller.setZoomLevel(level: CGFloat(level))
        result(success)
    }

    private func switchCamera(lens: String, result: @escaping FlutterResult) {
        guard let controller = cameraController else {
            result(Self.cameraError("NOT_INITIALIZED", "Camera not initialized"))
            return
        }

        controller.switchCamera(lens: lens) { state, error in
            DispatchQueue.main.async {
                if let error = error {
                    result(Self.cameraError("SWITCH_ERROR", error))
                } else {
                    if let dict = state as? [String: Any] {
                        DivineCameraLog.shared.info(
                            "Camera switched (lens=\(dict["lens"] ?? "?"))",
                            name: "DivineCamera.Lifecycle"
                        )
                    }
                    result(state)
                }
            }
        }
    }

    private func startRecording(
        maxDurationMs: Int?,
        useCache: Bool,
        outputDirectory: String?,
        audioDeviceId: String?,
        result: @escaping FlutterResult
    ) {
        guard let controller = cameraController else {
            result(Self.cameraError("NOT_INITIALIZED", "Camera not initialized"))
            return
        }

        controller.startRecording(
            maxDurationMs: maxDurationMs,
            useCache: useCache,
            outputDirectory: outputDirectory,
            audioDeviceId: audioDeviceId
        ) { error in
            DispatchQueue.main.async {
                if let error = error {
                    result(Self.cameraError("RECORD_START_ERROR", error))
                } else {
                    result(nil)
                }
            }
        }
    }

    private func stopRecording(result: @escaping FlutterResult) {
        guard let controller = cameraController else {
            result(Self.cameraError("NOT_INITIALIZED", "Camera not initialized"))
            return
        }

        controller.stopRecording { recordingResult, error in
            DispatchQueue.main.async {
                if let error = error {
                    result(Self.cameraError("RECORD_STOP_ERROR", error))
                } else {
                    result(recordingResult)
                }
            }
        }
    }

    private func pausePreview(result: @escaping FlutterResult) {
        cameraController?.pausePreview()
        result(nil)
    }

    private func resumePreview(result: @escaping FlutterResult) {
        cameraController?.resumePreview { state, error in
            DispatchQueue.main.async {
                if let error = error {
                    result(Self.cameraError("RESUME_ERROR", error))
                } else {
                    result(state)
                }
            }
        }
    }

    private func getCameraState(result: @escaping FlutterResult) {
        guard let controller = cameraController else {
            result(Self.cameraError("NOT_INITIALIZED", "Camera not initialized"))
            return
        }
        result(controller.getCameraState())
    }

    private func listAudioDevices(result: @escaping FlutterResult) {
        let discoverySession = AVCaptureDevice.DiscoverySession(
            deviceTypes: [.builtInMicrophone, .externalUnknown],
            mediaType: .audio,
            position: .unspecified
        )
        let devices: [[String: String]] = discoverySession.devices.map {
            device in
            [
                "id": device.uniqueID,
                "name": device.localizedName,
            ]
        }
        result(devices)
    }
}
